osu.Game/Online/HubClientConnector.cs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using osu.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Logging;
+using osu.Game.Online.API;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A component that maintains over a hub connection between client and server.
+    /// </summary>
+    public class HubClientConnector : Component
```

Comment:
Does this *really* need to be a component? I'd try implementing without it being one for simplicity (and avoiding needing to make the clients `CompositeDrawable`s).

Reply:
Made it a `Component` for simplicity of changes, but yeah actually now that you mention it, it doesn't really need to be a component. (given that it also make clients `CompositeDrawable` as you mentioned)

Reply:
How does it make things simpler? It's only using DI for one bindable (from the parent class, where it can just as easily be parsed in) and is not using the load or update flows. I'd argue that the way it uses BDL to perform setup is harder to parse and review than otherwise may be (because it creates potential cross-thread dependencies).